File: python/demo17a_reentrant_lock.py
'''
REENTRANT LOCKS (RECURSIVE MUTEXES)
Version A: Introduction to reentrant locks
'''

import threading



lock = threading.Lock()



def do_task():
    with lock:
        print('First time acquiring the resource')
        with lock:
            print('Second time acquiring the resource')



th = threading.Thread(target=do_task)
th.start()

# The thread th shall meet deadlock.
# So, you will never get output "Second time the acquiring resource".

File: cpp/cpp-boost/demo17a-reentrant-lock.cpp
/*
REENTRANT LOCKS (RECURSIVE MUTEXES)
Version A: Introduction to reentrant locks
*/


#include <iostream>
#include <boost/thread.hpp>
using namespace std;



boost::mutex mut;



void doTask() {
    mut.lock();
    cout << "First time acquiring the resource" << endl;

    mut.lock();
    cout << "Second time acquiring the resource" << endl;

    mut.unlock();
    mut.unlock();
}



int main() {
    boost::thread th(&doTask);
    /*
    The thread th shall meet deadlock.
    So, you will never get output "Second time the acquiring resource".
    */
    th.join();
    return 0;
}
